### PR TITLE
snap, snappy: implement the new (minmimal) kernel spec

### PR DESCRIPTION
**Pull request by mvo5**
This branch removes the current kernel specific yaml and replaces it with the new minimal kernel.yaml spec. This spec will evolve but for now I think we can get away with this. Hardcoding vmlinuz should not break anything, its used as the filename in pi2, dragonboard, i386 and amd64 already.

### CODE REVIEW

File: snap/kernel.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package snap
+
+import (
+	"fmt"
+
+	"gopkg.in/yaml.v2"
+)
+
+// KernelYaml contains the kernel.yaml specific data
+type KernelYaml struct {
```
Comment (pedronis) [inline snap/kernel.go]: later if this grows complicated will want to split kernelYaml vs KernelInfo but this seems cheap for now until things are more settled

File: snappy/kernel_os.go [hunk 2]
```diff
@@ -68,6 +68,15 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
 		return fmt.Errorf("can not extract kernel assets from snap type %q", s.Type)
 	}
 
+	// sanity check that we have the new kernel format
+	kmeta, err := snapf.ReadFile("meta/kernel.yaml")
```
Comment (pedronis) [inline snappy/kernel_os.go]: as we discussed snapf is probably not needed here (we already mounted the snap at this point) which would simplify things, also may be that we don't Unpack in snap.File (may still be good to have around until we really know more)

Comment (pedronis) [inline snappy/kernel_os.go]: from the point of view a snap.ReadKernelInfo() taking a Info or PlaceInfo and doing validation may make sense

File: snap/kernel.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	Version string `yaml:"version,omitempty"`
```
Comment (zyga) [inline snap/kernel.go]: Is this spec version or kernel version?

File: snap/kernel_test.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/ubuntu-core/snappy/snap"
+)
+
+type KernelYamlTestSuite struct {
+}
+
+var _ = Suite(&KernelYamlTestSuite{})
+
+func (s *KernelYamlTestSuite) TestValidateKernelYaml(c *C) {
+	err := snap.ValidateKernelYaml([]byte(`version: 4.4-18`))
```
Comment (zyga) [inline snap/kernel_test.go]: Ah, disregard my earlier comment

**Comment by zyga** [comment]
Looks good

File: snap/kernel.go (hunk 1 above)
Comment (niemeyer) [inline snap/kernel.go]: Can we please do that change now?  Right now it's very cheap to put the foundation in place as we want it. Once it grows legs, it becomes more boring.

Comment (niemeyer) [inline snap/kernel.go]: Ah, sorry, it was already done. Thanks!

File: snap/kernel.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+type kernelYaml struct {
+	Version string `yaml:"version,omitempty"`
+}
+
+type KernelInfo struct {
+	Version string
+}
+
+func ReadKernelInfo(info *Info) (*KernelInfo, error) {
```
Comment (niemeyer) [inline snap/kernel.go]: Some suggestions for error reporting for the function:

`const errorFormat = "cannot parse kernel snap details: %s"`

File: snap/kernel.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	kernelYamlFn := filepath.Join(info.MountDir(), "meta", "kernel.yaml")
+	kmeta, err := ioutil.ReadFile(kernelYamlFn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read %q: %s", kernelYamlFn, err)
```
Comment (niemeyer) [inline snap/kernel.go]: `return nil, fmt.Errorf(errorFormat, err)`

File: snap/kernel.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	var ky kernelYaml
+	if err := yaml.Unmarshal(kmeta, &ky); err != nil {
+		return nil, fmt.Errorf("info failed to parse: %s", err)
```
Comment (niemeyer) [inline snap/kernel.go]: `return nil, fmt.Errorf(errorFormat, err)`

File: snap/kernel.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	// basic validation
+	if ky.Version == "" {
+		return nil, fmt.Errorf("missing kernel version in kernel.yaml")
```
Comment (niemeyer) [inline snap/kernel.go]: `return nil, fmt.Errorf(errorFormat, "missing version in kernel.yaml")`

File: snappy/undo_test.go [hunk 9]
```diff
@@ -86,10 +86,11 @@ func (s *undoTestSuite) TestUndoForSetupSnapKernelUboot(c *C) {
 	}
 
 	testFiles := [][]string{
-		{"vmlinuz-4.4.0-14-generic.efi.signed", "kernel"},
-		{"initrd.img-4.4.0-14-generic", "initrd"},
+		{"vmlinuz", "kernel"},
+		{"initrd.img", "initrd"},
 		{"lib/modules/4.4.0-14-generic/foo.ko", "a module"},
```
Comment (niemeyer) [inline snappy/undo_test.go]: Shouldn't that be "modules/..." rather than "lib/modules/..."?

File: snappy/undo_test.go [hunk 10]
```diff
@@ -86,10 +86,11 @@ func (s *undoTestSuite) TestUndoForSetupSnapKernelUboot(c *C) {
...
 		{"lib/firmware/bar.bin", "some firmware"},
```
Comment (niemeyer) [inline snappy/undo_test.go]: Same.

File: snappy/kernel_os.go [hunk 11]
```diff
@@ -89,26 +97,24 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
...
-		if src == "" {
-			continue
-		}
-		if err := snapf.Unpack(src, dstDir); err != nil {
-			return err
-		}
-		src = filepath.Join(dstDir, src)
-		dst := filepath.Join(dstDir, dropVersionSuffix(src))
-		if err := os.Rename(src, dst); err != nil {
+	// TODO: hardcoded names until this is fully speced
```
Comment (niemeyer) [inline snappy/kernel_os.go]: I'd drop the comment, on the basis that this is actually our specification ATM.

File: snappy/kernel_os.go [hunk 12]
```diff
@@ -89,26 +97,24 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
...
 		if err := dir.Sync(); err != nil {
 			return err
 		}
 	}
-	if s.Legacy.Dtbs != "" {
-		src := filepath.Join(s.Legacy.Dtbs, "*")
-		dst := dstDir
-		if err := snapf.Unpack(src, dst); err != nil {
+
+	// TODO: hardcoded names until this is fully speced
```
Comment (niemeyer) [inline snappy/kernel_os.go]: Same, except the spec misses this entry. Let's talk about it.

File: snappy/kernel_os.go [hunk 13]
```diff
@@ -60,6 +55,13 @@ func removeKernelAssets(s snap.PlaceInfo, inter interacter) error {
 	return nil
 }
 
+func copyAll(src, dst string) error {
+	if output, err := exec.Command("cp", "-a", src, dst).CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to copy %q -> %q: %s (%q)", src, dst, err, output)
```
Comment (niemeyer) [inline snappy/kernel_os.go]: `"cannot copy %q to %q: %s (%s)"`

File: snappy/kernel_os.go [hunk 14]
```diff
@@ -68,6 +70,12 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
 		return fmt.Errorf("can not extract kernel assets from snap type %q", s.Type)
```
Comment (niemeyer) [inline snappy/kernel_os.go]: `s/can not/cannot/` here and below, please, since we're here

File: snap/kernel.go (hunk 5 above)
Comment (niemeyer) [inline snap/kernel.go]: (replaced "parse" by "read" on the message)

File: snap/kernel.go [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	var ky kernelYaml
+	if err := yaml.Unmarshal(kmeta, &ky); err != nil {
+		return nil, fmt.Errorf(errorFormat, err)
+	}
+
+	// basic validation
+	if ky.Version == "" {
+		return nil, fmt.Errorf(errorFormat, err)
```
Comment (pedronis) [inline snap/kernel.go]: probably quick copy pasting, there's no err here,  Gustavo suggested: return nil, fmt.Errorf(errorFormat, "missing version in kernel.yaml")

**Comment by mvo5** [comment]
@niemeyer Thanks for the detailed review. I addressed the points. The only open question is the `dtbs` directory which is not in the spec. Happy to talk about this.

File: snap/kernel.go (hunk 15 above)
Comment (mvo5) [inline snap/kernel.go]: Thank you!

**Comment by niemeyer** [comment]
retest this please

File: snap/kernel.go [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+type kernelYaml struct {
+	Version string `yaml:"version,omitempty"`
+}
+
+type KernelInfo struct {
+	Version string
+}
+
+const errorFormat = "cannot parse kernel snap details: %s"
```
Comment (niemeyer) [inline snap/kernel.go]: This should be inside the function rather than being package-wide, but we can fix that later.

**Comment by niemeyer** [comment]
LGTM

**Comment by pedronis** [comment]
retest this please

File: snappy/kernel_os.go [hunk 17]
```diff
@@ -89,26 +97,22 @@ func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inte
...
-			continue
-		}
-		if err := snapf.Unpack(src, dstDir); err != nil {
-			return err
-		}
-		src = filepath.Join(dstDir, src)
-		dst := filepath.Join(dstDir, dropVersionSuffix(src))
-		if err := os.Rename(src, dst); err != nil {
+	for _, src := range []string{
+		filepath.Join(s.MountDir(), "vmlinuz"),
```
Comment (pedronis) [inline snappy/kernel_os.go]: if I understood the discussion yesterday and look at the doc it seems this one now is named kernel.img

**Pull request merged**